File: src/main/java/org/raku/comma/psi/stub/RakuSubCallStubElementType.java
package org.raku.comma.psi.stub;

import com.intellij.psi.stubs.*;
import org.raku.comma.RakuLanguage;
import org.raku.comma.extensions.RakuFrameworkCall;
import org.raku.comma.psi.RakuSubCall;
import org.raku.comma.psi.impl.RakuSubCallImpl;
import org.raku.comma.psi.stub.impl.RakuSubCallStubImpl;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

public class RakuSubCallStubElementType extends IStubElementType<RakuSubCallStub, RakuSubCall> {
    public RakuSubCallStubElementType() {
        super("SUBCALL", RakuLanguage.INSTANCE);
    }

    @Override
    public RakuSubCall createPsi(@NotNull RakuSubCallStub stub) {
        return new RakuSubCallImpl(stub, this);
    }

    @NotNull
    @Override
    public RakuSubCallStub createStub(@NotNull RakuSubCall call, StubElement parentStub) {
        RakuFrameworkCall[] extensions = RakuFrameworkCall.EP_NAME.getExtensions();
        String calleeName = call.getCallName();
        Map<String, String> frameworkData = new HashMap<>();
        for (RakuFrameworkCall ext : extensions) {
            if (ext.isApplicable(call)) {
                String name = ext.getFrameworkName();
                for (Map.Entry<String, String> entry : ext.getFrameworkData(call).entrySet()) {
                    frameworkData.put(name + "." + entry.getKey(), entry.getValue());
                }
            }
        }
        return new RakuSubCallStubImpl(parentStub, calleeName, frameworkData);
    }

    @NotNull
    @Override
    public String getExternalId() {
        return "raku.stub.subcall";
    }

    @Override
    public void serialize(@NotNull RakuSubCallStub stub, @NotNull StubOutputStream dataStream) throws IOException {
        dataStream.writeName(stub.getName());
        Map<String, String> frameworkData = stub.getAllFrameworkData();
        dataStream.writeInt(frameworkData.size());
        for (Map.Entry<String, String> data : frameworkData.entrySet()) {
            dataStream.writeName(data.getKey());
            dataStream.writeUTF(data.getValue());
        }
    }

    @NotNull
    @Override
    public RakuSubCallStub deserialize(@NotNull StubInputStream dataStream, StubElement parentStub) throws IOException {
        String name = Objects.requireNonNull(dataStream.readName()).getString();
        Map<String, String> frameworkData = new HashMap<>();
        int elements = dataStream.readInt();
        for (int i = 0; i < elements; i++) {
            String key = Objects.requireNonNull(dataStream.readName()).getString();
            String value = dataStream.readUTF();
            frameworkData.put(key, value);
        }
        return new RakuSubCallStubImpl(parentStub, name, frameworkData);
    }

    @Override
    public void indexStub(@NotNull RakuSubCallStub stub, @NotNull IndexSink sink) {
        RakuFrameworkCall[] extensions = RakuFrameworkCall.EP_NAME.getExtensions();
        Map<String, String> allFrameworkData = stub.getAllFrameworkData();
        for (RakuFrameworkCall ext : extensions) {
            String prefix = ext.getFrameworkName();
            Map<String, String> frameworkData = new HashMap<>();
            for (Map.Entry<String, String> entry : allFrameworkData.entrySet()) {
                if (entry.getKey().startsWith(prefix + ".")) {
                    frameworkData.put(entry.getKey().substring(prefix.length() + 1), entry.getValue());
                }
            }
            if (! frameworkData.isEmpty()) {
                ext.indexStub(stub, frameworkData, sink);
            }
        }
    }
}

File: src/main/java/org/raku/comma/psi/stub/RakuSubCallStubElementType.kt
package org.raku.comma.psi.stub

import com.intellij.psi.stubs.*
import com.intellij.util.io.StringRef
import org.raku.comma.RakuLanguage
import org.raku.comma.extensions.RakuFrameworkCall
import org.raku.comma.psi.RakuSubCall
import org.raku.comma.psi.impl.RakuSubCallImpl
import org.raku.comma.psi.stub.impl.RakuSubCallStubImpl
import java.io.IOException
import java.util.*

class RakuSubCallStubElementType : IStubElementType<RakuSubCallStub?, RakuSubCall?>("SUBCALL", RakuLanguage.INSTANCE) {
    override fun createPsi(stub: RakuSubCallStub): RakuSubCall {
        return RakuSubCallImpl(stub, this)
    }

    override fun createStub(call: RakuSubCall, parentStub: StubElement<*>?): RakuSubCallStub {
        val extensions = RakuFrameworkCall.EP_NAME.extensions
        val calleeName = call.getCallName()
        val frameworkData: MutableMap<String?, String?> = HashMap<String?, String?>()
        for (ext in extensions) {
            if (ext.isApplicable(call)) {
                val name: String = ext.frameworkName
                for (entry in ext.getFrameworkData(call).entries) {
                    frameworkData.put(name + "." + entry.key, entry.value)
                }
            }
        }
        return RakuSubCallStubImpl(parentStub, calleeName, frameworkData)
    }

    override fun getExternalId(): String {
        return "raku.stub.subcall"
    }

    @Throws(IOException::class)
    override fun serialize(stub: RakuSubCallStub, dataStream: StubOutputStream) {
        dataStream.writeName(stub.getName())
        val frameworkData = stub.getAllFrameworkData()
        dataStream.writeInt(frameworkData.size)
        for (data in frameworkData.entries) {
            dataStream.writeName(data.key)
            dataStream.writeUTF(data.value)
        }
    }

    @Throws(IOException::class)
    override fun deserialize(dataStream: StubInputStream, parentStub: StubElement<*>?): RakuSubCallStub {
        val name = Objects.requireNonNull<StringRef?>(dataStream.readName()).getString()
        val frameworkData: MutableMap<String?, String?> = HashMap<String?, String?>()
        val elements = dataStream.readInt()
        for (i in 0..<elements) {
            val key = Objects.requireNonNull<StringRef?>(dataStream.readName()).getString()
            val value = dataStream.readUTF()
            frameworkData.put(key, value)
        }
        return RakuSubCallStubImpl(parentStub, name, frameworkData)
    }

    override fun indexStub(stub: RakuSubCallStub, sink: IndexSink) {
        val extensions = RakuFrameworkCall.EP_NAME.extensions
        val allFrameworkData = stub.getAllFrameworkData()
        for (ext in extensions) {
            val prefix = ext.frameworkName
            val frameworkData: MutableMap<String, String> = mutableMapOf()
            for (entry in allFrameworkData.entries) {
                if (entry.key!!.startsWith("$prefix.")) {
                    frameworkData.put(entry.key!!.substring(prefix.length + 1), entry.value)
                }
            }
            if (!frameworkData.isEmpty()) {
                ext.indexStub(stub, frameworkData, sink)
            }
        }
    }
}
